examples/Custom_Hooks/campaign_stopping.py

```diff
@@ -0,0 +1,261 @@
+}
+dict_ligand = {
+    "BrettPhos": r"CC(C)C1=CC(C(C)C)=C(C(C(C)C)=C1)C2=C(P(C3CCCCC3)C4CCCCC4)C(OC)="
+    "CC=C2OC",
+    "Di-tert-butylphenylphosphine": r"CC(C)(C)P(C1=CC=CC=C1)C(C)(C)C",
+    "(t-Bu)PhCPhos": r"CN(C)C1=CC=CC(N(C)C)=C1C2=CC=CC=C2P(C(C)(C)C)C3=CC=CC=C3",
+    "Tricyclohexylphosphine": r"P(C1CCCCC1)(C2CCCCC2)C3CCCCC3",
+    "PPh3": r"P(C1=CC=CC=C1)(C2=CC=CC=C2)C3=CC=CC=C3",
+    "XPhos": r"CC(C1=C(C2=CC=CC=C2P(C3CCCCC3)C4CCCCC4)C(C(C)C)=CC(C(C)C)=C1)C",
+    "P(2-furyl)3": r"P(C1=CC=CO1)(C2=CC=CO2)C3=CC=CO3",
+    "Methyldiphenylphosphine": r"CP(C1=CC=CC=C1)C2=CC=CC=C2",
+    "1268824-69-6": r"CC(OC1=C(P(C2CCCCC2)C3CCCCC3)C(OC(C)C)=CC=C1)C",
+    "JackiePhos": r"FC(F)(F)C1=CC(P(C2=C(C3=C(C(C)C)C=C(C(C)C)C=C3C(C)C)C(OC)=CC=C2OC)"
+    r"C4=CC(C(F)(F)F)=CC(C(F)(F)F)=C4)=CC(C(F)(F)F)=C1",
+    "SCHEMBL15068049": r"C[C@]1(O2)O[C@](C[C@]2(C)P3C4=CC=CC=C4)(C)O[C@]3(C)C1",
+    "Me2PPh": r"CP(C)C1=CC=CC=C1",
+}
+
+parameters = [
+    SubstanceParameter(name="Solvent", data=dict_solvent, encoding="MORDRED"),
+    SubstanceParameter(name="Base", data=dict_base, encoding="MORDRED"),
+    SubstanceParameter(name="Ligand", data=dict_ligand, encoding="MORDRED"),
+    NumericalDiscreteParameter(name="Temp_C", values=[90, 105, 120], tolerance=2),
+    NumericalDiscreteParameter(name="Concentration", values=[0.057, 0.1, 0.153]),
+]
+
+searchspace = SearchSpace.from_product(parameters=parameters)
+
+objective = SingleTargetObjective(target=NumericalTarget(name="yield", mode="MAX"))
+
+recommender = TwoPhaseMetaRecommender(
+    initial_recommender=RandomRecommender(), recommender=BotorchRecommender()
+)
+
+# Also, we load the dataframe containing the lookup data for the closed-loop simulation:
+
+try:
+    lookup = pd.read_excel("./../Backtesting/lookup.xlsx")
+except FileNotFoundError:
+    lookup = pd.read_excel("examples/Backtesting/lookup.xlsx")
+
+### Simulating the Unstopped Campaigns
+
+# First, we run several Monte Carlo repetitions of the uninterrupted campaign to get a
+# feeling for the average trajectory:
+
+campaign = Campaign(searchspace, objective, recommender)
+results_unstopped = simulate_scenarios(
+    {"Average uninterrupted": campaign},
+    lookup,
+    batch_size=BATCH_SIZE,
+    n_doe_iterations=N_DOE_ITERATIONS,
+    n_mc_iterations=N_MC_ITERATIONS,
+)
+
+
+### Defining the Campaign-Stopping Hook
+
+# In order to interrupt a running campaign, we define a custom exception to identify our
+# stopping event:
+
+
+class CampaignStoppedException(Exception):
+    """The campaign should be stopped."""
+
+
+# Based on this exception class, we can now define a hook implementing the stopping
+# criterion. For this purpose, we count the fraction of candidates with a PI exceeding a
+# given value and terminate the campaign once the fraction falls below a certain
+# threshold.
+
+PI_THRESHOLD = 0.01  # PI of 1% to identify promising points
+PI_REQUIRED_FRACTION = 0.1  # 10% of candidates must be above the threshold
```

Comment:
Hi @Scienfitz, I've reorganized the example quite a bit. Apart from revising the comments: both stopped and unstopped simulation now use the same logic, and bits of code have been moved to more appropriate sections. I think it should be easier for a new user to follow / compare. Let me know if you see issues.

However, there is one big problem I see. The number here don't work for me at all. That is, the campaigns are never stopped early! I don't know how you've been able to create the plots with this setting tbh. Note that this is not a problem coming from my changes, but it exists already for your original version! (BTW: I've created a backup of the unmodified branch under `backup/dev/campaing_stopping` so that you can verify. But in the new branch, I deleted the images because they were added in several commits and – as I said – I'm not even able to reproduce them).

Can you have a look and see if you can reproduce the problem?

Reply:
@AdrianSosic I believe the rebase destroyed the plots
the rebased branch had the updated mll fit (torch 100 steps), this one was originally on the scipy one

this has also happened for rims picture, it is not reproduced...

i have tested a bit and increasing steps improves the plots but this is a dilemma

Reply:
Dang!!!! What now?

Reply:
* try and see whether an over larger step size for torch fit works
* could have an if clause that select scipy fit always except if theres a taskparameter present then its doing torch100
* could have an environment variable controlling the fit

Reply:
Alright, second option seems like the most sensible temporary workaround to me. Let me set up a mini-PR for that, then I'll do a local test merge to verify if both examples work again. Then we can merge this branch and you can run your test suite from last time again before we release

Reply:
Perhaps, in the meantime while I'm doing the first part, you can read once again over the refactored example and see if everything is OK